File: frontend/src/types/rental.ts
/**
 * Rental Management Types (Story 4.3)
 */

export type RentalStatus = "active" | "overdue" | "returned";
export type ReturnCondition = "good" | "damaged" | "lost";

export interface RentalListItem {
  order_item_id: string;
  garment_id: string;
  garment_name: string;
  customer_name: string;
  customer_phone: string;
  start_date: string; // ISO date
  end_date: string; // ISO date
  rental_days: number;
  days_remaining: number;
  rental_status: RentalStatus;
  deposit_amount: string; // Decimal as string
  unit_price: string; // Decimal as string
  image_url: string | null;
}

export interface RentalListParams {
  status?: RentalStatus;
  search?: string;
  sort_by?: "end_date" | "days_remaining" | "customer_name";
  page?: number;
  page_size?: number;
}

export interface PaginationMeta {
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface RentalListResponse {
  data: RentalListItem[];
  meta: {
    pagination: PaginationMeta;
  };
}

export interface RentalStats {
  active_rentals: number;
  overdue_rentals: number;
  due_this_week: number;
  returned_this_month: number;
}

export interface ProcessReturnInput {
  return_condition: ReturnCondition;
  damage_notes?: string;
  deposit_deduction: string; // Decimal as string
}

export interface ProcessReturnResponse {
  order_item_id: string;
  garment_id: string;
  return_condition: ReturnCondition;
  deposit_deduction: string;
  damage_notes: string | null;
  returned_at: string; // ISO datetime
}

export interface RentalReturnHistory {
  return_condition: ReturnCondition;
  damage_notes: string | null;
  returned_at: string;
}

export interface RentalDetailResponse {
  order_item_id: string;
  garment_id: string;
  garment_name: string;
  image_url: string | null;
  category: string;
  customer_name: string;
  customer_phone: string;
  customer_email: string | null;
  start_date: string;
  end_date: string;
  rental_days: number;
  days_remaining: number;
  rental_status: RentalStatus;
  deposit_amount: string;
  unit_price: string;
  order_id: string;
  return_history: RentalReturnHistory | null;
}

File: backend/src/models/rental.py
"""Pydantic schemas for Rental Management (Story 4.3)."""

from datetime import date, datetime
from decimal import Decimal
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, Field


class RentalStatus(str, Enum):
    """Rental item status."""

    active = "active"
    overdue = "overdue"
    returned = "returned"


class ReturnCondition(str, Enum):
    """Condition of returned rental item."""

    good = "good"
    damaged = "damaged"
    lost = "lost"


class RentalListItem(BaseModel):
    """Schema for a rental item in list response."""

    order_item_id: UUID
    garment_id: UUID
    garment_name: str
    customer_name: str
    customer_phone: str
    start_date: date
    end_date: date
    rental_days: int
    days_remaining: int  # Computed: end_date - today
    rental_status: str  # 'active', 'overdue', 'returned'
    deposit_amount: Decimal
    unit_price: Decimal
    image_url: str | None = None

    model_config = {"from_attributes": True}


class RentalListParams(BaseModel):
    """Filter parameters for rental list."""

    status: str | None = Field(None, pattern=r"^(active|overdue|returned)$")
    search: str | None = Field(None, max_length=255)  # Search by garment name, customer name, or phone
    sort_by: str = Field("end_date", pattern=r"^(end_date|days_remaining|customer_name)$")
    page: int = Field(1, ge=1)
    page_size: int = Field(20, ge=1, le=100)


class RentalListResponse(BaseModel):
    """Response for rental list endpoint."""

    data: list[RentalListItem]
    meta: dict = Field(
        default_factory=dict,
        description="Pagination metadata: {pagination: {total, page, page_size, total_pages}}"
    )


class RentalStats(BaseModel):
    """Summary statistics for rental dashboard."""

    active_rentals: int = Field(..., description="Dang thue count")
    overdue_rentals: int = Field(..., description="Qua han count")
    due_this_week: int = Field(..., description="Tra trong tuan count")
    returned_this_month: int = Field(..., description="Da tra thang nay count")


class ProcessReturnInput(BaseModel):
    """Schema for processing return of rental item."""

    return_condition: ReturnCondition
    damage_notes: str | None = Field(None, max_length=500)
    deposit_deduction: Decimal = Field(..., ge=0)


class ProcessReturnResponse(BaseModel):
    """Response after processing return."""

    order_item_id: UUID
    garment_id: UUID
    return_condition: str
    deposit_deduction: Decimal
    damage_notes: str | None = None
    returned_at: datetime

    model_config = {"from_attributes": True}


class RentalDetailResponse(BaseModel):
    """Detail view of a single rental item."""

    order_item_id: UUID
    garment_id: UUID
    garment_name: str
    image_url: str | None = None
    category: str
    customer_name: str
    customer_phone: str
    customer_email: str | None = None
    start_date: date
    end_date: date
    rental_days: int
    days_remaining: int
    rental_status: str
    deposit_amount: Decimal
    unit_price: Decimal
    order_id: UUID
    return_history: dict | None = None  # Latest return if exists

    model_config = {"from_attributes": True}
